tests/test_auto_sharding_fwd_only.py

```diff
@@ -0,0 +1,77 @@
+import unittest
+import os
+
+import jax
+import alpa
+import ray
+import numpy as np
+from jax import numpy as jnp
+from jax import lax
+from alpa.pipeline_parallel.primitive_def import mark_pipeline, mark_gradient
+from alpa import automatic_layer_construction
+from alpa.util import get_ray_namespace_str
+
+
+class AutoShardingFwdOnlyTest(unittest.TestCase):
+    def setUp(self):
+        os.environ["XLA_PYTHON_CLIENT_ALLOCATOR"] = "platform"
+        assert len(jax.local_devices()) >= 4
+
+        ray.init(address="auto",
+                 namespace=get_ray_namespace_str(prefix="alpa-unittest"))
+
+    def tearDown(self):
+        ray.shutdown()
+
+    def test_pipeshard_fwd_only(self):
+        device_mesh = alpa.DeviceCluster().get_virtual_physical_mesh()
+        alpa.set_parallelize_options(
+            devices=device_mesh,
+            strategy="pipeshard_parallel",
+            pipeline_stage_mode="auto_gpipe",
+            num_micro_batches=2,
+        )
+
+        def func(x):
+            bufs_updated = {}
+            bufs_updated["_dummy"] = jnp.ones(1)
+            x = jnp.matmul(x, x)
+            x = jnp.matmul(x, x)
+            y = x + 1
+            return bufs_updated, (x, y)
+
+        def wrapper_func(x):
```

Comment:
This is very much a hack because I wasn't able to get fwd-only to work without having a apply_grad stage (XLA would just return empty protos). So here I add a dummy apply_grad stage that satisfies these properties: 1. all outputs are tracked, 2. the dummy operations in the apply_grad stage are simple so that it doesn't take up much compute.